src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs

```diff
@@ -0,0 +1,314 @@
+            {
+                lock (this) // protect _isntrumentConfigFuncs list
+                {
+                    foreach (var filter in _instrumentConfigFuncs)
+                    {
+                        if (filter(instrument))
+                        {
+                            instrumentState = BuildInstrumentState(instrument);
+                            if (instrumentState != null)
+                            {
+                                _instrumentStates.TryAdd(instrument, instrumentState);
+                                instrumentState = _instrumentStates[instrument];
+                            }
+                            break;
+                        }
+                    }
+                }
+            }
+            return instrumentState;
+        }
+
+        internal InstrumentState? BuildInstrumentState(Instrument instrument)
+        {
+            Func<Aggregator>? createAggregatorFunc = GetAggregatorFactory(instrument);
+            if (createAggregatorFunc == null)
+            {
+                return null;
+            }
+            Type aggregatorType = createAggregatorFunc.GetType().GenericTypeArguments[0];
+            Type instrumentStateType = typeof(InstrumentState<>).MakeGenericType(aggregatorType);
+            return (InstrumentState)Activator.CreateInstance(instrumentStateType, createAggregatorFunc)!;
+        }
+
+        private static Func<Aggregator>? GetAggregatorFactory(Instrument instrument)
+        {
+            Type type = instrument.GetType();
+            Type? genericDefType = null;
+#if IS_GENERIC_TYPE_SUPPORT
+            genericDefType = type.IsGenericType ? type.GetGenericTypeDefinition() : null;
+#else
+            try
+            {
+                 genericDefType = type.GetGenericTypeDefinition();
+            }
+            catch {}
+#endif
+            if (genericDefType == typeof(Counter<>))
+            {
+                return () => new RateSumAggregator();
+            }
+            else if (genericDefType == typeof(ObservableCounter<>))
+            {
+                return () => new RateAggregator();
+            }
+            else if (genericDefType == typeof(ObservableGauge<>))
+            {
+                return () => new LastValue();
+            }
+            else if (genericDefType == typeof(Histogram<>))
+            {
+                return () => new ExponentialHistogramAggregator(DefaultHistogramConfig);
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        internal void Collect()
+        {
+            _listener.RecordObservableInstruments();
+
+            foreach (KeyValuePair<Instrument, InstrumentState> kv in _instrumentStates)
+            {
+                kv.Value.Collect(kv.Key, (LabeledAggregationStatistics labeledAggStats) =>
+                {
+                    _collectMeasurement(kv.Key, labeledAggStats);
+                });
```

Comment:
How frequently does this code run / how much do we care about overheads?  This, for example, is allocating a new delegate and a new closure for every pair in _instrumentStates.

Reply:
In a typical scenario Collect() runs between once per second and once per 10 minutes. In the limit it could run 10 times/sec. Typical scenario probably has 10-100 items in this loop. 

Thus far I paid very little care to allocations in this path with the expectation the scenario impact would be neglible.